src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+// At least 15% different than the shortest path.
+const constexpr auto kGamma = 0.85;
+// Alternative paths are still reasonable around the via node candidate (local optimality).
+// At least optimal around 10% sub-paths around the via node candidate.
+const /*constexpr*/ auto kAlpha = 0.10;
+// Note: ^ ICEs gcc 7.1, just leave it out for better times
+
+// Represents a via middle node where forward (from s) and backward (from t)
+// search spaces overlap and the weight a path (made up of s,via and via,t) has.
+struct WeightedViaNode
+{
+    NodeID node;
+    EdgeWeight weight;
+};
+
+// Represents a complete packed path (made up of s,via and via,t)
+// its total weight and the via node used to construct the path.
+struct WeightedViaNodePackedPath
+{
+    WeightedViaNode via;
+    PackedPath path;
+};
+
+// Represents a high-detail unpacked path (s, .., via, .., t)
+// its total weight and the via node used to construct the path.
+struct WeightedViaNodeUnpackedPath
+{
+    WeightedViaNode via;
+    UnpackedNodes nodes;
+    UnpackedEdges edges;
+};
+
+// Filters candidates which are on not unique.
+// Returns an iterator to the uniquified range's new end.
+// Note: mutates the range in-place invalidating iterators.
+template <typename RandIt> RandIt filterViaCandidatesByUniqueNodeIds(RandIt first, RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    std::sort(first, last, [](auto lhs, auto rhs) { return lhs.node < rhs.node; });
+    return std::unique(first, last, [](auto lhs, auto rhs) { return lhs.node == rhs.node; });
+}
+
+// Filters candidates which are on un-important roads.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterViaCandidatesByRoadImportance(RandIt first, RandIt last, const Facade &facade)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    // Todo: the idea here is to filter out alternatives where the via candidate is not on a
+    // high-priority road. We should experiment if this is really needed or if the boundary
+    // nodes the mld search space gives us already provides us with reasonable via candidates.
+    //
+    // Implementation: we already have `RoadClassification` from guidance. We need to serialize
+    // it to disk and then in the facades read it in again providing `IsImportantRoad(NodeID)`.
+    // Note: serialize out bit vector keyed by node id with 0/1 <=> unimportant/important.
+    (void)first;
+    (void)last;
+    (void)facade;
+
+    return last;
+}
+
+// Filters candidates with much higher weight than the primary route. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterViaCandidatesByStretch(RandIt first, RandIt last, EdgeWeight weight)
```

Comment:
this kinda indicates to me that the stretch-test would be OK after this test. We test a global stretch, but we want to test the local stretch in the end. The total weight is only a very bad indicator:

Lets take `a --- b --- c  --- d` and `a --- b --- e --- d`, `|abcd| = 1`, |abed| = (1+0.25)`.

If `|bcd| == 0.8 * |abcd|`, the local stretch of `bed` vs `bcd` is `((1+0.25) - 0.2) / 0.8 = 1.31`, for  `|bcd| == 0.5 * |abcd|` you already get a stretch of `1.5`.  In total your stretch gets scaled here by the amount of sharing, which is an undesired effect.